### ISSUE TITLE
Code Climate Integration

### ISSUE THREAD

**Issue by akroon3r** [BCDevOps/devops-requests#397]
## Step 0
**Do you have repos on bcgov GitHub Organization?**
If yes, please proceed with the request.

**If there is similar applications/tools available in the community, have you tried them out?**
There are many well supported community tools for various purposes, such SonarCube and BDDStack. Before you start with the new tools, have you consider trying out those that are available and widely adopted by the other application teams? If you have done so and do found that the existing ones does not match your need, please proceed.

Code Climate refines Sonar Qube functionality to add a hard quality gate into the development lifecycle, removing Technical Debt introduced in development before code is merged into master.

**Have you tested out this GitHub application already?**
We understand that there are a lot of new and cool GitHub Applications out there that teams could like to try out. However, we cannot just enable the integration without evaluating the app, as those are open to all application teams. We believe you have tested out the GitHub apps and had good reasoning for the preference before making this request.

example repo: [NTT Data Notification Service](https://github.com/SierraSystems/ntt-data-notification-service)

## Step 1
Please fill out the team and project info so we know how to contact you:

* Product Name: JAG File Submission Application
* Team Name: AG/PSSG - NTTData 
* GitHub Repositories on bcgov: [JAG File Submission](https://github.com/bcgov/jag-file-submission)

## Step 2
Please provide details about the requested GitHub App. Take a look here for a (good example)[https://github.com/BCDevOps/platform-services/issues/120/]

1. Brief introduction of the app:

Code Climate is a third party, github approved application that utilizes Sonar Qube to report on code quality metrics. Integrating this application into a repo allows code climate to report directly on a Pull Request, creating a git action that can fully block any addition of code to a particular branch if pre-defined code quality thresholds are not met. The application will create messages on the code diff created, pointing out to the submitter where said violations occur. This application will greatly aid in the removal of technical debt before an application moves into a DEV, TEST or PROD environment.

2. General Criteria:
  - How Easy / Fast / Repeatable is it to use and maintain for both Development Teams and Platform Services Teams: 
Development teams interact with the application on the level of the Pull Request, and view reported metrics by clicking on generated badges which can be viewed on the repositories README. Platform service teams set the application up once by following provided documentation, and let developers interact with the application once they begin submitting code. On both accounts it is easy, fast and repeatable to use.
  - Customizable:
Code Climate allows for specific configurations to be set through a .codeclimate.yml file which can be included in the root level of a repository. Here you control metrics like argument count, complex log, file length, identical blocks of code, method complexity, method count, method count, method length, nested control flow, return statements and similar blocks of code. Also, Code climate allows for controlling the % threshold of test coverage on a PR. Essentially ensuring full functionality of Sonar Qube but fully integrated with Github.
  - User Access Model:
Repository Admins have read and write access to the Code Climate application, whereas normal users only have read access.
  - License:
[https://github.com/codeclimate/codeclimate/blob/master/LICENSE](https://github.com/codeclimate/codeclimate/blob/master/LICENSE)
  - BCGov STRA:
@NickCorcoran said in our initial disocvery meeting that his assessment will run in parallel with our pilot, if approved.
Privacy Policy -> https://codeclimate.com/legal/privacy
  - Price Model:
Free for public repositories

3. Comparison between Existing Tools Used by the Community
  - pros:
Reporting, github integration, code quality gate creation, moving technical debt reduction to a different stage of the development lifecycle
Basic discovery of Sonar Cloud has shown that it does not support Java projects out of the box (which we do a majority of our development in), whereas Code Climate has the ability to support Java out of the box.
Initial setup time for Code Climate was ~1h, whereas fully integrating with Sonar in openshift can take upwards of 4-8 hours based on your level of experience integrating with Jenkins pipelines etc.
  - cons:
You're still using Sonar Qube, just integrated to a different stage of the development lifecycle, so the only real con would be that developers would have to deal with a ramp up of using a different tool that utilizes the same platform (sonar qube).
- potential issues
Code Climate doesn't retain issue history, whereas the current image of Sonar Qube available to the Openshift 3 platform can display this to users (we haven't found out if this is actually the case, perhaps a configuration we haven't found or used)

4. Relevant Resources and Links
[Github Marketplace](https://github.com/marketplace/code-climate)
[Code Climate Docs](https://docs.codeclimate.com/)

**Comment by ShellyXueHan** [comment]
@akroon3r I've enabled code climate for your repo, please let us know if you have issue accessing it.

**Comment by alexjoybc** [comment]
Thanks a lot @ShellyXueHan for setting this up for us, our team will start configuring code climate tomorrow, we will let you know how it goes ;-)

**Comment by akroon3r** [comment]
WIP - https://github.com/bcgov/jag-file-submission 

You check see our badges on the project readme and I'm working on getting the appropriate actions setup this morning!

**Comment by akroon3r** [comment]
Test coverage report is up :)

**Comment by akroon3r** [comment]
https://codeclimate.com/github/bcgov/jag-file-submission

We've got the application reporting issues and test coverage successfully. Had to get through a lot of issues WRT to reporting on an application that has a frontend and backend in the same repo.

The git action can be seen in the /.github/workflows/ directory. 

If you have any questions, please let us know. I'm working on documenting next and hopefully a demo at some point!

**Comment by akroon3r** [comment]
We've run into one issue:

Forked repositories that submit code via PR need access to the CC_TEST_REPORTER_ID secret, and as Github doesn't allow for secrets to be shared between origin and forks, this creates an issue.

So far cloning the repository and submitting code via PR doesn't run into this issue. For the time being we're working from clones instead of forks as a work around.

**Comment by akroon3r** [comment]
misclicked sorry!

**Comment by ShellyXueHan** [comment]
> We've run into one issue:
> 
> Forked repositories that submit code via PR need access to the CC_TEST_REPORTER_ID secret, and as Github doesn't allow for secrets to be shared between origin and forks, this creates an issue.
> 
> So far cloning the repository and submitting code via PR doesn't run into this issue. For the time being we're working from clones instead of forks as a work around.

Could you also include this in the pros/cons session? Or maybe track it as `Issue`? Is this a known thing for codeclimate?

**Comment by akroon3r** [comment]
Will do!

**Comment by akroon3r** [comment]
https://github.com/bcgov/jag-file-submission/issues/205

**Comment by caggles** [comment]
Hey! Is this ticket still active? What's up with this one now?
